### PR TITLE
Taking serial number from 'show inventory' instead of 'show version'

### PR DESCRIPTION
**Pull request by jjagodzinski**
Serial number from 'show version' gives SN of processor. Serial number
from 'show inventory' gives you serial numbers of chassis, power unit
and other subdevices. This fix returns only serial number for chassis.
It is possible to retrieve information about other modules from device.
Change is needed because on ASA5545 SN of chassis and processor might
be diffrent. We use chassis SN to recognize devices.

### CODE REVIEW

File: src/ralph/scan/plugins/ssh_cisco_asa.py [hunk 1]
```diff
@@ -104,12 +104,24 @@ def scan_address(ip_address, **kwargs):
...
     try:
         lines = ssh.asa_command(
-            "show version | grep (^Hardware|Boot microcode|^Serial|address is)"
+            "show version | grep (^Hardware|Boot microcode|address is)"
         )
+        lines2 = ssh.asa_command("show inventory")
     finally:
         ssh.close()
+
+    inventory = [item for item in lines2 if item != u'\r']
```
Comment (mkurek) [inline src/ralph/scan/plugins/ssh_cisco_asa.py]: there is `from __future__ import unicode_literals` at the top of the file, so you don't have to use `u` before string:
`u'\r'` -> `'\r'`

File: src/ralph/scan/plugins/ssh_cisco_asa.py [hunk 2]
```diff
@@ -104,12 +104,24 @@ def scan_address(ip_address, **kwargs):
...
+    components = []
+    for parts in [i + j for i, j in zip(inventory[::2], inventory[1::2])]:
+        element = {}
+        for item in re.findall(r'(?:[^,,"]|"(?:\\.|[^"])*")+',
+                               re.sub('\r', ',', parts)):
+            k, v = item.split(":")
+            element[k.strip().lower().replace('"',
```
Comment (mkurek) [inline src/ralph/scan/plugins/ssh_cisco_asa.py]: couldn't you replace `"` in whole string (item) at once?

File: src/ralph/scan/plugins/ssh_cisco_asa.py [hunk 3]
```diff
@@ -104,12 +104,24 @@ def scan_address(ip_address, **kwargs):
...
         )
+        lines2 = ssh.asa_command("show inventory")
     finally:
         ssh.close()
+
+    inventory = [item for item in lines2 if item != u'\r']
+    components = []
+    for parts in [i + j for i, j in zip(inventory[::2], inventory[1::2])]:
+        element = {}
+        for item in re.findall(r'(?:[^,,"]|"(?:\\.|[^"])*")+',
```
Comment (ghost) [inline src/ralph/scan/plugins/ssh_cisco_asa.py]: This is so unreadable, IMHO. Please give a hint here at least, by named var or comment.

File: src/ralph/scan/plugins/ssh_cisco_asa.py (hunk 2 above)
Comment (ghost) [inline src/ralph/scan/plugins/ssh_cisco_asa.py]: Maybe it's me ;) but this:
`k.strip().lower().replace('"','')`
also could be assigned to descriptive variable.

File: src/ralph/scan/plugins/ssh_cisco_asa.py (hunk 1 above)
Comment (jjagodzinski) [inline src/ralph/scan/plugins/ssh_cisco_asa.py]: changed

File: src/ralph/scan/plugins/ssh_cisco_asa.py (hunk 3 above)
Comment (jjagodzinski) [inline src/ralph/scan/plugins/ssh_cisco_asa.py]: added comment

File: src/ralph/scan/plugins/ssh_cisco_asa.py [hunk 4]
```diff
@@ -104,12 +104,28 @@ def scan_address(ip_address, **kwargs):
...
+    #             u'pid': u'ASA-PWR-AC', u'sn': u'XXXX', u'vid': u'N/A'}, ...]
+    inventory = [item for item in lines2 if item != '\r']
+    subdevices = []
+    for subdevice in [i + ", " + j for i, j in zip(inventory[::2],
+                                                   inventory[1::2])]:
+        parameters = {}
+        for parameter in re.findall(r'(?:[^,"]|"(?:\.|[^"])*")+', subdevice):
+            name, value = parameter.split(":")
+            parameters[name.strip().lower()] = value.strip().replace('"', '')
+        subdevices.append(parameters)
```
Comment (ghost) [inline src/ralph/scan/plugins/ssh_cisco_asa.py:130]: This should tested, but it's a way more readible, imho :)

```
inventory = [item for item in lines2 if item != '\r']
subdevices = []
def gen_subdevices(inventory):
    even_lines = inventory[::2]
    odd_lines = inventory[1::2]
    for even_line, odd_line in zip(even_lines, odd_lines):
        subdevice = ', '.join(even_line, odd_line)
        yield subdevice
# TODO: add descriptive comment - plain english - for this regexp
pair_finder = re.compile(r'(?:[^,"]|"(?:\.|[^"])*")+')
for subdevice in gen_subdevice(inventory):
    parameters = {}
    for parameter in pair_finder.findall(subdevice):
        name, value = parameter.split(":")
        parameters[name.strip().lower()] = value.strip().replace('"', '')
    subdevices.append(parameters)
```

**Comment by mkurek** [comment]
@jjagodzinski do you still want to merge it? If yes, please apply @xliiv improvements.

**Comment by mkurek** [comment]
:ok:

**Comment by ghost** [comment]
:ok:

**Pull request merged**